File: 2023/4.py
input = open(f"{__file__.replace('py', 'input')}").readlines()

total_sum = 0
cards = [1] * len(input)
for i, line in enumerate(input):
    line = line.split(": ")[1]
    my, winning = line.split(" | ")
    my = my.split()
    winning = winning.split()
    common = len(set(my) & set(winning))

    if common > 0:
        total_sum += pow(2, common - 1)

        for j in range(common):
            cards[i + j + 1] += cards[i]

print(total_sum)
print(sum(cards))

File: 2023/4.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input'))
input = input.split("\n").map { _1.split(': ')[1] }

sum = 0
cards = input.size.times.map { 1 }
input.each_with_index do |line, i|
  my = line.split(' | ')[0].split.map(&:to_i)
  winning = line.split(' | ')[1].split.map(&:to_i)

  common = (my & winning).size

  next if common.zero?

  sum += (2**(common - 1))
  common.times do |j|
    cards[i + j + 1] += cards[i]
  end
end

p sum
p cards.sum
